collector/textfile.go

```diff
@@ -52,31 +65,165 @@ func NewTextFileCollector() (Collector, error) {
 	c := &textFileCollector{
 		path: *textFileDirectory,
 	}
+	return c, nil
+}
 
-	if c.path == "" {
-		// This collector is enabled by default, so do not fail if
-		// the flag is not passed.
-		log.Infof("No directory specified, see --collector.textfile.directory")
-	} else {
-		textFileAddOnce.Do(func() {
-			prometheus.DefaultGatherer = prometheus.Gatherers{
-				prometheus.DefaultGatherer,
-				prometheus.GathererFunc(func() ([]*dto.MetricFamily, error) { return c.parseTextFiles(), nil }),
+func convertMetricFamilies(metricFamilies []*dto.MetricFamily, ch chan<- prometheus.Metric) {
+	var valType prometheus.ValueType
+	var val float64
+
+	for _, mf := range metricFamilies {
+		labelNames := map[string]struct{}{}
+		for _, metric := range mf.Metric {
+			labelPairs := metric.GetLabel()
+			for _, label := range labelPairs {
+				if _, ok := labelNames[label.GetName()]; !ok {
+					labelNames[label.GetName()] = struct{}{}
+				}
 			}
-		})
+		}
+
+		for _, metric := range mf.Metric {
+			labelPairs := metric.GetLabel()
+			var labels []string
+			var labelVals []string
+			for _, label := range labelPairs {
+				labels = append(labels, label.GetName())
+				labelVals = append(labelVals, label.GetValue())
+			}
+
+			for k := range labelNames {
+				present := false
+				for _, label := range labels {
+					if k == label {
+						present = true
+						break
+					}
+				}
+				if present == false {
+					labels = append(labels, k)
+					labelVals = append(labelVals, "")
+				}
+			}
+
+			metricType := mf.GetType()
+			switch metricType {
+			case dto.MetricType_COUNTER:
+				if metric.Counter != nil {
+					valType = prometheus.CounterValue
+					val = metric.Counter.GetValue()
+				}
+			case dto.MetricType_GAUGE:
+				if metric.Gauge != nil {
+					valType = prometheus.GaugeValue
+					val = metric.Gauge.GetValue()
+				}
+			case dto.MetricType_UNTYPED:
+				if metric.Untyped != nil {
+					valType = prometheus.UntypedValue
+					val = metric.Untyped.GetValue()
+				}
+			case dto.MetricType_SUMMARY:
+				if metric.Summary != nil {
+					quantiles := map[float64]float64{}
+					for _, q := range metric.Summary.Quantile {
+						quantiles[q.GetQuantile()] = q.GetValue()
+					}
+					ch <- prometheus.MustNewConstSummary(
+						prometheus.NewDesc(
+							*mf.Name,
+							mf.GetHelp(),
+							labels, nil,
+						),
+						metric.Summary.GetSampleCount(),
+						metric.Summary.GetSampleSum(),
+						quantiles, labelVals...,
+					)
+				}
+			case dto.MetricType_HISTOGRAM:
+				if metric.Histogram != nil {
+					buckets := map[float64]uint64{}
+					for _, b := range metric.Histogram.Bucket {
+						buckets[b.GetUpperBound()] = b.GetCumulativeCount()
+					}
+					ch <- prometheus.MustNewConstHistogram(
+						prometheus.NewDesc(
+							*mf.Name,
+							mf.GetHelp(),
+							labels, nil,
+						),
+						metric.Histogram.GetSampleCount(),
+						metric.Histogram.GetSampleSum(),
+						buckets, labelVals...,
+					)
+				}
+
```

Comment:
There's an extra empty line here.

To catch programming errors, let's add a `default` case here that does `panic("unknown metric type")`.